Exclude empty chats (with no ADDCOMMENT actions or message drafts) from LNH

PauloGasparSv opened the pull request:
### Details

Updating the LNH logic so it doesn't show empty chats! We consider a chat empty if it's not active and has no drafts or messages (ADDCOMMENT actions).
This change must not influence how Threads, IOUS, Tasks and other features of the App work.

### Fixed Issues
$ https://github.com/Expensify/App/issues/14523
PROPOSAL: N/A


### Tests
 

#### Empty chat shouldn't stay in LNH

1. Sign into newDot
2. Start a new chat with an account.
3. Make sure the new chat opens and is shown in the LNH
4. [WEB ONLY]: Copy the chat's `reportID` (last number in the chat's URL)
5. Open another existing chat (like concierge), make sure the chat disappears from the LNH
6. Start a chat again with that same account, make sure the account shows up in the Options list since you already had a chat with it before
7. Make sure the chat opens and has the same `reportID` (has the same URL)


#### Chat should stay in LNH if it has a Draft

1. In the same chat of used in the previous test, type some text into the composer but do not send the message
2. Open another existing chat (like concierge), make sure the chat stays in the LNH and shows the draft icon

#### Not-Empty chat should stay in LNH

1. Send a message in the same chat from the previous test.
2. Open another existing chat (like concierge), make sure the chat stays in the LNH.

#### Deleting messages and leaving chat empty should remove it from LNH

1. Open a chat that has a message sent inside it
2. Delete that message so the chat becomes empty
3. Open another existing chat (like concierge), make sure the chat disappears from the LNH.

#### Same tests w/ Group Chat 

1. Start a new group chat with two accounts "B" and "C"
2. Make sure the new chat opens and is shown in the LNH
3. [WEB ONLY]: Copy the chat's `reportID` (last number in the chat's URL)
4. Open another existing chat (like concierge), make sure the chat disappears from the LNH
5. Start a group chat again with account B and C
6. Make sure the chat opens and has the same `reportID` (has the same URL)

#### Money request should stay in LNH 

1. Start a new empty chat as userA with existing account userB
2. Sign in as userB and send a message back to userA so you become known
3. As userA, send new Money Request to userB inside the chat
4. As userB, delete the message sent in step 2 so the chat becomes empty
5. As userA, Open another existing chat (like concierge), make sure the reports don't disappear from the LNH
6. Make sure both the Money Request and the original report stay in the LNH

#### Workspace Chats should stay in LNH
1. Create a new Workspace in an account
2. Make sure the `#admins`, `#announce` and a workspace chats are created
6. Open another existing chat (like concierge), make sure none of the workspace related chats disappear from the LNH
4. Invite another user to the workspace, make sure another workspace chat is created for the invited user
5. Open another existing chat (like concierge), make sure none of the workspace related chats disappear from the LNH


TODO: add steps for these

- [ ] Verify that no errors appear in the JS console

### Offline tests

1. Sign into newDot
2. Turn connectivity off
3. Start a new chat with an account.
4. Make sure the new chat opens and is shown in the LNH
5. [WEB ONLY]: Copy that chat's `reportID` (last number in the chat's URL)
6. Open another existing chat (like concierge), make sure the chat disappears from the LNH
7. Start a chat again with that same account, make sure the account shows up in the Options list since you already had a chat with it before
8. [WEB ONLY]: That chat's `reportID` will not be the same as the first attempt because we are offline, but make note of that `reportID` as well
9. Open another existing chat (like concierge), make sure the chat disappears from the LNH
10. [WEB ONLY]: Toggle connectivity back on
11. [WEB ONLY]: Start a chat again with that same account
12. [WEB ONLY]: Make sure that chat's `reportID` is the same as the from step 5 (and not step 8)

 
Make sure the chat opens and has the same `reportID` (has the same URL)


### QA Steps
Same as Tests

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>


https://github.com/Expensify/App/assets/6564265/48a56b70-4826-4aa5-bbf7-6f5e4baba99c

https://github.com/Expensify/App/assets/6564265/9f2e4ac6-4360-4649-b612-5e25f89fdb64


</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/6564265/f0baf5d8-9b32-4f2b-9f39-5d152f6431fc


https://github.com/Expensify/App/assets/6564265/3344beb2-d256-485f-b1ef-d578b0ba7bc4


<!-- add screenshots or videos here -->

</details>

<details>
<summary>Mobile Web - Safari</summary>


https://github.com/Expensify/App/assets/6564265/feff3f6f-de26-4a96-b07d-0e06ef81c3d5



</details>

<details>
<summary>Desktop</summary>


https://github.com/Expensify/App/assets/6564265/dd0d767c-9505-439c-8c5b-226bd6ee847c


</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/6564265/5437335a-c650-4d6a-8614-2682c81d9227


</details>

<details>
<summary>Android</summary>


https://github.com/Expensify/App/assets/6564265/2256d885-d1ba-4e16-973d-7059257ab28f



</details>

Commit:
Exclude empty chat reports from LNH

Commit:
Fix typo in isChatReport call

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-exclude-empty-chats-from-lnh

Commit:
Add empty chat LNH test to SidebarFilterTest

PauloGasparSv commented:
Having some trouble fixing the Travis errors in `SideBarOrderTest` and `SideBarFilterTest`, getting back to this tomorrow.

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-exclude-empty-chats-from-lnh

Commit:
Add excludeEmptyChats logic to shouldReportBeInOptionList

Commit:
Exclude threads, default rooms and policy rooms from empty chat logic in shouldReportBeInOptionList

Commit:
Add another test to SidebarFilterTest for reports with drafts that should render in LNH

Commit:
Fix policy expense beta test in SidebarFilterTest

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-exclude-empty-chats-from-lnh

Commit:
Linting

Commit:
Fix empty chat with draft LNH test

Commit:
Add lastMessageText to SideBarOrderTest reports so they show in LNH

Commit:
Change LNH empty chat check order

PauloGasparSv commented:
I think I fixed all tests!

Since the `SidebarOrderTest` tests deal with ordering and checking the created/updated time of actions, instead of creating a comment inside the test so the reports stick to the LNH I mocked the `lastMessageHTML` field on the reports. This way I don't influence the last action time or anything related to those tests

PauloGasparSv commented:
I'm trying to understand and fix the following bug before un-drafting this:

1. Start a new chat and send 2 messages
2. Delete them
3. Go to another chat, the chat should disappear from the LNH but it takes some time before it does that

https://github.com/Expensify/App/assets/6564265/e84b8271-e0fd-41a5-8e93-89676954f930

Commit:
Do not include groups in logic that excludes empty chats from LNH

Commit:
Fix group chat check for excluding empty chats from newDot

Commit:
Lint ReportUtils

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-exclude-empty-chats-from-lnh

Commit:
Exclude empty group chats from the LNH

PauloGasparSv commented:
I think I still have a problem to fix:

Sending a message then going to another chat really quickly makes the Chat disappear then pop back in! I think there might be some optimistic data in the chat creation that I should be setting to avoid that.

Will fix this tomorrow

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-exclude-empty-chats-from-lnh

PauloGasparSv commented:
> 
> I think I still have a problem to fix:
> 
> Sending a message then going to another chat really quickly makes the Chat disappear then pop back in! I think there might be some optimistic data in the chat creation that I should be setting to avoid that.
> 
> Will fix this tomorrow
> 
> 

Trying to reproduce this reliably to fix it, this is not always happening!

I managed to reproduce this a couple of times by deleting multiple comments inside the report and then going to the concierge chat really quickly.

I think we are not clearing `lastMessageText` somewhere correctly and that is generating the behaviour.

Still trying to understand how force that behavior to replicate this 

<details>

![image](https://github.com/Expensify/App/assets/6564265/4e87adf4-e38a-48d2-aa80-9a8812e10b15)


https://github.com/Expensify/App/assets/6564265/cd57b85d-70a0-4071-b9bb-10fbb6d12137

</details>

PauloGasparSv commented:
I can't reproduce [this](https://github.com/Expensify/App/pull/19321#issuecomment-1579316306) reliably and eventually the app fixes itself when that happens. Since it's an edge-case that can only happen if we delete the last messages of a chat and exit I won't focus on it anymore.

I think it won't be a problem but if someone reports that we can work on it separately from here.

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-exclude-empty-chats-from-lnh

PauloGasparSv commented:
After merging [this](https://expensify.slack.com/archives/C03TQ48KC/p1686087481767219) into `main` the behavior changed, the active chat is also being excluded from LNH.

Trying to fix that.

Besides that, I was having some trouble testing IOU's, they were failing in my VM and I'm debugging the reason here.

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-exclude-empty-chats-from-lnh

PauloGasparSv commented:
Right now I'm trying to understand the logic changes [that happened here](https://github.com/Expensify/App/pull/19321#issuecomment-1579493082) so this logic works again.

Commit:
Lint ReportUtils

PauloGasparSv commented:
I think [this check](https://github.com/Expensify/App/blob/d7c45a2e8e776f345f566f0cf1ae6d08f7c2e8a0/src/libs/ReportUtils.js#L1862) that checks if the report is currently opened is broken and that's why the feature isn't working anymore.

The `reportIDFromRoute` is empty in my tests, investigating why here!
<img width="878" alt="image" src="https://github.com/Expensify/App/assets/6564265/8d905f23-2642-4c65-97d9-aa9797e3151f">

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-exclude-empty-chats-from-lnh

grgia commented:
I encountered the [same bug](https://github.com/Expensify/App/pull/19321#issuecomment-1590140672), brought it up here- https://expensify.slack.com/archives/C04878MDF34/p1686218695956109

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-exclude-empty-chats-from-lnh

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-exclude-empty-chats-from-lnh

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-exclude-empty-chats-from-lnh

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-exclude-empty-chats-from-lnh

Commit:
Fixing tests by using lastMessageText instead of lastMessageHtml

Commit:
Fix new sidebarFilterTest tests

Commit:
Lint tests

PauloGasparSv commented:
Ok, I think we are almost done here!

I found 2 issues to investigate now regarding optimisticData and connectivity:
1. When we create a Money Request to a user we never chatted with before we create a new Report for the request and also create a new action in the user's Chat report. I'm my tests there is a small period of time before all responses are sent where the Chat Report can disappear from the LNH.
I think that is happening because of OptimisticData we are not using correctly, still need to investigate further.

2. When we go offline and create a new Chat with someone newDot generates a new random `reportID`. That is a problem because if the user doesn't type anything inside the chat and exit it then try to re-create the chat a new `reportID` will be used and once the user is online then a bunch of different `OpenReport` calls will be made for different `reportID`s but for the same 2 participants.
I'm not sure how much of a problem that is, still need to investigate further.

Will shift focus and go back to this tomorrow.

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-exclude-empty-chats-from-lnh

PauloGasparSv commented:
The bug #2 I commented above really is a problem, the API creates a new chat pretty every time we ask it to : /

<img width="410" alt="image" src="https://github.com/Expensify/App/assets/6564265/b02d1fac-e9f0-4ed0-9037-82a62e71f5dc">
(Opening a chat with a new Account while offline 3 times and removing the exclude chat from LNH logic)

I'll fix bug #1 first and then go back to thinking on how to fix #2

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-exclude-empty-chats-from-lnh

PauloGasparSv commented:
Going back to this rn:

I can replicate error no. 2 while online too : /
Noticed it while testing item 1

https://github.com/Expensify/App/assets/6564265/0c5e50fd-e61b-4e69-8c14-70d6e407245c

Commit:
Replace getChatByParticipantsByLoginList call with getChatByParticipants in navigateToAndOpenReport

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-exclude-empty-chats-from-lnh

Commit:
Undo previous change in navigateToAndOpenReport and call navigateToAndOpenReportWithAccountIDs instead

PauloGasparSv commented:
I fixed the new errors where the report was created over and over again with a new reportID while online (still have the other errors including doing this while offline).

|'m still not sure why this fix was required so I'm comparing with main right now, I added a call to `navigateToAndOpenReportWithAccountIDs` in `createChat` in case the options have an  `accountID` instead of using `navigateToAndOpenReport` and the logins, this broke creating chats with new users so I need to fix this if I go with this.

Once I'm done with that I'll go back to the offline error

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-exclude-empty-chats-from-lnh

Commit:
Reintroduce excludeEmptyChats logic

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-exclude-empty-chats-from-lnh

Commit:
Reintroduce navigateToAndOpenReport login navigation

luacmartins commented on src/libs/ReportUtils.js:
```diff
@@ -2398,6 +2399,16 @@ function shouldReportBeInOptionList(report, currentReportId, isInGSDMode, iouRep
     }
 
+    // Hide thread reports that haven't been commented on
+    if (isThread(report) && _.isEmpty(report.lastMessageText)) {
```
I think it's better to use [getLastVisibleMessage](https://github.com/Expensify/App/blob/65ee88e55afcdd0f60db506c95b20e6ccb2139f9/src/libs/ReportActionsUtils.js#L384), since it'll loop over the report actions, in case the report object missed an update. We should also check for the draft comment here

```suggestion
    if (isThread(report) && _.isEmpty(ReportActionsUtils.getLastVisibleMessage(report)['lastMessageText']) && report.hasDraftComment) {
```

PauloGasparSv replied:
Makes sense! Thks for the suggestion and already putting the `lastMessageText` check in the snippet since that function can also return the created action : )

Commit:
Use getLastVisibleMessage to check for empty chats in shouldReportBeInOptionList

Commit:
Reuse isEmptyChat for empty chat threads check in shouldReportBeInOptionList

Commit:
Fix typo and lint

Commit:
Remove duplicated and faulty logic that hides empty threads

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-exclude-empty-chats-from-lnh

Commit:
Fix SidebarOrderTest tests

PauloGasparSv commented:
Hey @thesahindia @luacmartins this is ready for reviews!

I'm adding the missing evidences right now but please feel free to test : )

Commit:
Keep empty policy expense chats in LNH

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-exclude-empty-chats-from-lnh

PauloGasparSv commented:
I'm having trouble with the Android emulator so I'll add it's evidence tomorrow morning

Commit:
Merge branch 'main' of github.com:Expensify/App into paulogasparsv-exclude-empty-chats-from-lnh

PauloGasparSv commented:
I think I'll be missing Android evidence [because of this internal error](https://expensify.slack.com/archives/C05F0U613LY/p1691621785891289) we are trying to solve.

Checking it of the author's checklist!

luacmartins reviewed (approved):
LGTM

luacmartins commented:
@thesahindia bump for review!

thesahindia reviewed (approved):
Tested well! 

P.S. Sorry for the delay. I am having internet issues.

luacmartins commented:
Not sure why Joel got assigned since we already had an internal engineer assigned. Merging.

joh42 commented:
Running into a failing unit test on main, seems to be from this PR:
<img width="877" alt="Screenshot 2023-08-13 at 15 09 10" src="https://github.com/Expensify/App/assets/138504087/49409fda-b515-43b0-8b40-bc8008eded9a">
@luacmartins @PauloGasparSv

Beamanator commented on tests/unit/SidebarFilterTest.js:
```diff
@@ -71,7 +72,59 @@ describe('Sidebar', () => {
+                        }),
+                    )
+
+                    // Then no reports are rendered in the LHN
```
Probably copy / pasta error? Should be "The report is rendered in the LHN"

PauloGasparSv replied:
Oh yes, thks for catching that!

I'll create a P.R. to fix that unit test that is failing + fix that comment.

PauloGasparSv commented:
> Running into a failing unit test on main, seems to be from this PR:

Thks for tagging me and [reporting here](https://expensify.slack.com/archives/C049HHMV9SM/p1691934002284919), taking a look at that test right now : )

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5859813282) to staging by https://github.com/luacmartins in version: 1.3.54-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|failure ❌ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5859813282) to staging by https://github.com/luacmartins in version: 1.3.54-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|failure ❌ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5885771241) to production by https://github.com/yuwenmemon in version: 1.3.54-13 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
